doc: add Doxygen preview tooltips to C symbols

kartben opened the pull request:
https://builds.zephyrproject.io/zephyr/pr/77093/docs/kernel/services/threads/index.html#thread-termination

<img width="500" alt="image" src="https://github.com/user-attachments/assets/3f004394-5434-4729-a0a6-5ec6a96150ed">

Add Doxygen tooltips in the main documentation.

The tooltips are triggered when hovering over C symbols in the documentation and display a preview of the Doxygen documentation for the symbol.

Some shadow DOM magic is needed to style the tooltips correctly, without affecting the rest of the page.

Commit from #77095 is needed to prevent compliance error due to clang trying to run on JS file and failing.

License of the tooltip library is MIT - not sure we need a formal approval to use it (other bits of the website use e.g. CC-BY-SA) (@nashif?)

Will likely extend on this in a follow-up PR to also add tooltips to Kconfig options links.

faxe1008 commented:
Not relevant for this PR but maybe worth wile 😀: for public methods have a "jump to source" link. In some instances I needed to check what assumptions a method makes about the passed in parameters, which is not always evident by looking at the docs.

I know that this is tricky since most of the time function definition and implementation is in separate files, but maybe there is a smart way to so this?

pdgendt commented:
Is there a way to clear previous compliance comments? They appear to stack.

kartben committed:
doc: extensions: Add Doxygen tooltips to main documentation

Add Doxygen tooltips in the main documentation.

The tooltips are triggered when hovering over C symbols in the
documentation and display a preview of the Doxygen documentation for the
symbol.

Some shadow DOM magic is needed to style the tooltips correctly, without
affecting the rest of the page.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

gmarull commented on doc/_static/js/doxygen-tooltip.js:
```diff
@@ -0,0 +1,129 @@
+
+  let style = document.createElement('style');
+
+  style.textContent += `.tippy-box {`;
```
is there any reason to not serve this in a regular css file?

kartben replied:
good point, no particular reason besides that it eventually needs to be attached to the shadow root, not the main document. It's now served as a regular static resource but some gymnastics is needed to pass it to the shadow root

gmarull reviewed (commented):
would it be possible to make this a regular Sphinx extension that injects js/css? like mathjax, kconfig, code copy button etc.

kartben commented:
> would it be possible to make this a regular Sphinx extension that injects js/css? like mathjax, kconfig, code copy button etc.

sure, will do. Thanks for the review!

gmarull commented on doc/_extensions/zephyr/doxytooltip/static/doxytooltip.css:
```diff
@@ -0,0 +1,18 @@
+  overflow-y: scroll;
+  background-color: var(--content-wrap-background-color);
+  color: var(--body-color);
+}
```
missing nl

gmarull commented on doc/_extensions/zephyr/doxytooltip/__init__.py:
```diff
@@ -0,0 +1,36 @@
+"""
+Doxygen Tooltip Extension
+#################
```
align `###`

gmarull reviewed (commented):
lgtm, just nitpicks

gmarull commented:
IMHO, popup font is too small, or maybe it's me who needs a new pair of glasses.

![image](https://github.com/user-attachments/assets/3e1e487e-91c8-43ce-84bb-27875434d2b0)

pdgendt commented on doc/_extensions/zephyr/doxytooltip/static/tippy/popper.min.js:
```diff
@@ -0,0 +1,8 @@
+ */
+
+!function(e,t){"object"==typeof exports&&"undefined"!=typeof module?t(exports):"function"==typeof define&&define.amd?define(["exports"],t):t((e="undefined"!=typeof globalThis?globalThis:e||self).Popper={})}(this,(function(e){"use strict";function t(e){if(null==e)return window;if("[object Window]"!==e.toString()){var t=e.ownerDocument;return t&&t.defaultView||window}return e}function n(e){return e ...
+//# sourceMappingURL=popper.min.js.map
```
Do we also want the `.map` files to debug potential issues?

kartben replied:
I lean towards no (mostly because these are two 100KB+ files to add to the repo) but don't really have a strong opinion. @gmarull thoughts?

kartben commented:
> IMHO, popup font is too small, or maybe it's me who needs a new pair of glasses.

I've increased from 0.7em to 0.8em to buy you some time until you need them new glasses 🤓

I've dropped the commit from https://github.com/zephyrproject-rtos/zephyr/pull/77095 so compliance check failure is expected and a false positive (but the fact it *fails* is actually suspicious. maybe due to huge single lines on the minified JS files? @pdgendt)

pdgendt commented:
> I've dropped the commit from #77095 so compliance check failure is expected and a false positive (but the fact it _fails_ is actually suspicious. maybe due to huge single lines on the minified JS files? @pdgendt)

The issue is due to a missing newline at the end of the minified lines..

kartben commented:
* removed a 404 css link that was leftover from before moving to a sphinx extension
* fixed styling of the tooltip which was causing the tooltip arrows to not show up and scrollbars to flicker in some situations
* increased size to 0.9em :)

arrows now working:
<img width="98" alt="image" src="https://github.com/user-attachments/assets/66202717-ed2f-4070-b24b-d2e212cb2d88">

kartben commented:
> Not relevant for this PR but maybe worth wile 😀: for public methods have a "jump to source" link. In some instances I needed to check what assumptions a method makes about the passed in parameters, which is not always evident by looking at the docs.
> 
> I know that this is tricky since most of the time function definition and implementation is in separate files, but maybe there is a smart way to do this?

I am really not sure there is an easy way to be honest, plus, there might be several implementations available too, right? So not sure the user experience would be all that great. I'd rather we improve on the way the "assumptions" you describe are being documented :) Cheers!

nordicjm commented:
Very buggy on ios mobile, will show "loading" then just navigate to the information page most of the time, if you then go back and tap on it, it will show the mini information

kartben commented:
> Very buggy on ios mobile, will show "loading" then just navigate to the information page most of the time, if you then go back and tap on it, it will show the mini information

@nordicjm Should be better now, thanks for spotting this.
Also rebased to get recent fix in compliance check (clang format won't try to run on JS files)

nordicjm commented:
Seems like opening one (again, mobile ios) does not preclude opening another, if you open two right next to each other then one can obscure the other. And if you have only open when scrolling, it will flip position when it goes across the half way point in the page which then obscures the text on the page and makes the page jump

kartben commented:
I am tempted to just disable the tooltips on mobile then?

nashif commented:
to be on the safe side, let's bring the license topic into the TSC today.

kartben commented:
Hey @npitre - would be curious to get your feedback re: this new feature and whether it is in a good shape in terms of accessibility. Thanks!

henrikbrixandersen commented:
> I am tempted to just disable the tooltips on mobile then?

I think that's a good idea.

henrikbrixandersen commented:
> > I am tempted to just disable the tooltips on mobile then?
> 
> I think that's a good idea. Tooltips in general are not nice on (small) touchscreens.

kartben commented:
* disabled tooltips on "focus" event (i.e. won't show when using key navigation)
* disabled entirely on touch devices

kartben commented:
Just to clarify, as people have brought up that this could hinder accessibility: the latest update to this PR is effectively preventing the tooltips to show for people who are not navigating the website using their mouse, and they would just click the links as before to get to the API docs. 

I will probably be opening a separate issue to track accessibility improvement opportunities on the website (ex. improving contrast) but for the time being, I am confident this PR is not making things any worse than they were before.
